utils/polus-filepattern-util/filepattern/filepattern.py

```diff
@@ -0,0 +1,445 @@
+            for key, value in variables.items():
+                new_entry[key] = value
+        temp_dict.append(new_entry)
+
+    for key in uvals.keys():
+        uvals[key].sort()
+    
+    return file_ind, uvals
+
+def get_matching(files,var_order,out_var=None,**kwargs):
+    """ Get filenames that have defined variable values
+    
+    This gets all filenames that match a set of variable values. Variables must be one of
+    filename.VARIABLES, and the inputs must be uppercase. The following example code would
+    return all files that have c=0:
+    pattern = "file_x{xxx}_y{yyy}_c{ccc}.ome.tif"
+    file_path = "./path/to/files"
+    files = parse_directory(file_path,pattern,var_order='cyx')
+    channel_zero = get_matching(files,'cyx',C=0)
+
+    Multiple coordinates can be used simultaneously, so in addition to C=0 in the above example,
+    it is also possible to include Y=0. Further, each variable can be a list of values, and the
+    returned output will contain filenames matching any of the input values.
+
+    Inputs:
+        files - A file dictionary (see parse_directory)
+        var_order - A string indicating the order of variables in a nested output dictionary
+        out_var - Variable to store results, used for recursion
+        kwargs - One of filepatter.VARIABLES, must be uppercase, can be single values or a list of values
+    Outputs:
+        out_var - A list of all files matching the input values
+    """
+    # Initialize the output variable if needed
+    if out_var == None:
+        out_var = []
+        
+    # If there is no var_order, then files should be a list of files.
+    if len(var_order)==0:
+        if not isinstance(files,list):
+            TypeError('Expected files to be a list since var_order is empty.')
+        out_var.extend(files)
+        return
+
+    for arg in kwargs.keys():
+        assert arg==arg.upper() and arg.lower() in VARIABLES, "Input keyword arguments must be uppercase variables (one of R, T, C, Z, Y, X, P)"
+    
+    if var_order[0].upper() in kwargs.keys():
+        if isinstance(kwargs[var_order[0].upper()],list): # If input was already a list
+            v_iter = kwargs[var_order[0].upper()]
+        else:                                             # If input was not a list, make it a list
+            v_iter = [kwargs[var_order[0].upper()]]
+    else:
+        v_iter = [i for i in files.keys()]
+    v_iter.sort()
+    
+    for v_i in v_iter:
+        if v_i not in files.keys():
+            continue
+        get_matching(files[v_i],var_order[1:],out_var,**kwargs)
+    return out_var
+
+class FilePattern():
+    """ Main class for handling filename patterns
+    
+    Most of the functions in filepattern.py return complicated variable structures that might
+    be difficult to use in an abstract way. This class provides tools to use the above functions
+    more simple. In particular, the iterate function is an iterable that permits simple
```

Comment:
typo, this class provides tools to make the use of the above functions more simple